### PR TITLE
Renaming of the library and use requires_grad()

### PR DESCRIPTION
**Pull request by ceriottm**
Following our discussion, just did a quick & dirty implementation of full fw gradients

### CODE REVIEW

**Comment by frostedoyster** [comment]
I fixed the one against e3nn in the main branch

File: benchmarks/benchmark.py [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
 from torch.profiler import profile
 
 device = "cpu"
 
 l_max = 10
-sh_calculator = torch_sh.SphericalHarmonics(l_max, device)
+sh_calculator = sphericart_torch.SphericalHarmonics(l_max, device)
 
-xyz = torch.rand((10000, 3), device=device, requires_grad=True)
+xyz = torch.rand((1000000, 3), device=device, requires_grad=True)
```
Comment (frostedoyster) [inline benchmarks/benchmark.py:12]: We need these to be realistic. Fixing the benchmarks/profiling is a different problem

File: benchmarks/benchmark_e3nn.py [hunk 2]
```diff
@@ -10,7 +10,7 @@
 #yzx = torch.rand((100000, 3), device=device, requires_grad=False)
 #sh = e3nn.o3.spherical_harmonics(range(l_max+1), yzx, normalize=True, normalization='integral')
 
-yzx = torch.rand((10000, 3), device=device, requires_grad=True)
+yzx = torch.rand((1000000, 3), device=device, requires_grad=True)
```
Comment (frostedoyster) [inline benchmarks/benchmark_e3nn.py:13]: Same here

File: sphericart_torch/extension/spherical_harmonics.cc [hunk 3]
```diff
@@ -8,28 +8,32 @@ class SphericalHarmonics : public torch::autograd::Function<SphericalHarmonics>
...
             int n_samples = xyz.sizes()[0];
-            torch::Tensor spherical_harmonics = torch::empty({n_samples, (l_max+1)*(l_max+1)}, xyz.options());
-            double* dsph = nullptr;
-            cartesian_spherical_harmonics_generic(n_samples, l_max, prefactors.data_ptr<double>(), xyz.data_ptr<double>(), spherical_harmonics.data_ptr<double>(), dsph);
+            torch::Tensor spherical_harmonics = torch::empty({n_samples, (l_max+1)*(l_max+1)}, xyz.options());                
+            if (xyz.requires_grad()) {
+                torch::Tensor spherical_harmonics_gradients = torch::empty({n_samples, 3, (l_max+1)*(l_max+1)}, xyz.options());
+                cartesian_spherical_harmonics(n_samples, l_max, prefactors.data_ptr<double>(), 
+                    xyz.data_ptr<double>(), spherical_harmonics.data_ptr<double>(), spherical_harmonics_gradients.data_ptr<double>());
+                ctx->save_for_backward({xyz, spherical_harmonics_gradients});
```
Comment (frostedoyster) [inline sphericart_torch/extension/spherical_harmonics.cc:17]: If we're after raw speed, I would save the gradients inside saved_data instead. `save_for_backward` performs additional checks, like the in-place modification check, that are not needed if the tensor is not available to the user. Contrast this to xyz, for which the check is needed because the user could modify it outside our function.
That said, I don't know what the C++ syntax to make this change is. I would have to try digging into the C++ source files again, since the documentation is poor. In short, we can also leave it like this for the moment and then I'll fix it

File: sphericart_torch/extension/spherical_harmonics.cc [hunk 4]
```diff
@@ -8,28 +8,32 @@ class SphericalHarmonics : public torch::autograd::Function<SphericalHarmonics>
...
 
         static torch::autograd::tensor_list backward(torch::autograd::AutogradContext *ctx, torch::autograd::tensor_list d_loss_d_outputs) {
             int l_max = ctx->saved_data["l_max"].toInt();
             auto saved = ctx->get_saved_variables();
-            torch::Tensor prefactors = saved[0];
-            torch::Tensor xyz = saved[1];
+            torch::Tensor xyz = saved[0];
+            if (! xyz.requires_grad() ) {
+                throw "Cannot compute backwards, no derivatives have been computed";
+            }
```
Comment (frostedoyster) [inline sphericart_torch/extension/spherical_harmonics.cc:32]: I think we're safe here because the backward function will never be called if requires_grad is false, so we can take this out.

**Review by frostedoyster — COMMENTED** [review]
Perfect. This should be close to our final interface
